### PR TITLE
Use C#9 Covariant return types feature for Clone() implementation for type safety and reduce number of casts

### PR DESCRIPTION
**Pull request by SergeiPavlov**

### CODE REVIEW

File: Orm/Xtensive.Orm/Sql/Dml/Expressions/SqlUserFunctionCall.cs
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Sql/Dml/Expressions/SqlUserFunctionCall.cs:31]: This code is broken, i suppose it should be 
```suggestion
    internal override SqlUserFunctionCall Clone(SqlNodeCloneContext context) =>
      context.GetOrAdd(this, static (t, c) =>
        new SqlUserFunctionCall(t.Name, t.Arguments.Select(o => (SqlExpression) o.Clone(c)).ToArray(t.Arguments.Count)));
```

File: Orm/Xtensive.Orm/Sql/Dml/Expressions/SqlFunctionCall.cs
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Sql/Dml/Expressions/SqlFunctionCall.cs:31]: This code is broken, i suppose it should be 
```suggestion
    internal override SqlFunctionCall Clone(SqlNodeCloneContext context) =>
      context.GetOrAdd(this, static (t, c) =>
        new SqlFunctionCall(t.FunctionType, t.Arguments.Select(o => o.Clone(c)).ToArray(t.Arguments.Count)));
```

File: Orm/Xtensive.Orm/Sql/Dml/Expressions/SqlCustomFunctionCall.cs
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Sql/Dml/Expressions/SqlCustomFunctionCall.cs:34]: This code is broken, I suppose it should look like

```suggestion
    internal override SqlCustomFunctionCall Clone(SqlNodeCloneContext context) =>
      context.GetOrAdd(this, static (t, c) =>
        new SqlCustomFunctionCall(t.FunctionType, t.Arguments.Select(o => o.Clone(c)).ToArray(t.Arguments.Count)));
```

File: Orm/Xtensive.Orm/Sql/Dml/Hints/SqlForceJoinOrderHint.cs
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Sql/Dml/Hints/SqlForceJoinOrderHint.cs:23]: Should we do following
```
    new SqlForceJoinOrderHint((t.tables is null)
          ? Array.Empty<SqlTable>()
          : t.tables.Select(table => (SqlTable) table.Clone()).ToArray(t.tables.Length)));
```

I didn't add it as suggestion deliberately because ```using Xtensive.Core;``` has to be added to namespaces

**Comment by alex-kulakov** [comment]
Please, do not merge master to here before fixing building issues.

File: Orm/Xtensive.Orm/Sql/Dml/Hints/SqlForceJoinOrderHint.cs
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Sql/Dml/Hints/SqlForceJoinOrderHint.cs:23]: > Should we do following:

No. if the original `table == nul` then clone's `table` should be null, not empty array.

**Comment by SergeiPavlov** [comment]
> Please, do not merge master to here before fixing building issues.

THe build is fixed

File: Orm/Xtensive.Orm/Sql/Dml/Hints/SqlForceJoinOrderHint.cs
Comment (alex-kulakov) [reply Orm/Xtensive.Orm/Sql/Dml/Hints/SqlForceJoinOrderHint.cs:23]: Yes, you're right, I should've replaced Array.Empty with null
```
new SqlForceJoinOrderHint((t.tables is null)
          ? null
          : t.tables.Select(table => (SqlTable) table.Clone()).ToArray(t.tables.Length)));
```
I wanted to use .ToArray(lenght) extension.

**Comment by alex-kulakov** [comment]
> THe build is fixed

Nice. thank you

File: Orm/Xtensive.Orm/Sql/Dml/Hints/SqlForceJoinOrderHint.cs
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Sql/Dml/Hints/SqlForceJoinOrderHint.cs:23]: Actually it is not necessary
`.ToArray()` extension smart enough to extract `.Count()` of its source.Select().

This is the effective implementation:
(https://github.com/dotnet/runtime/blob/main/src/libraries/System.Linq/src/System/Linq/Select.SpeedOpt.cs#L74)

**Comment by alex-kulakov** [comment]
Error appeared on our build server after merge: Orm\Xtensive.Orm\Sql\Dml\Expressions\SqlExtract.cs(75,62): error CS8821: A static anonymous function cannot contain a reference to 'this' or 'base'.

**Comment by SergeiPavlov** [comment]
> Error appeared on our build server after merge: Orm\Xtensive.Orm\Sql\Dml\Expressions\SqlExtract.cs(75,62): error CS8821: A static anonymous function cannot contain a reference to 'this' or 'base'.

Thank you!
Fixed.

**Comment by alex-kulakov** [comment]
@SergeiPavlov, please, merge changes from master, I'll pull-request to st fork on top of this merge to fix minor issues with SqlTruncateTable cloning and make it ready to merge. Merge of SqlInsert will cause some problems.

**Review by alex-kulakov — CHANGES_REQUESTED** [review]
Everything seems to be good except for new node - SqlTruncateTable, which has old cloning which breaks building. I would also add line between field and property in the SqlNodeCloneContext, I suggested this change for you to be easy to apply.

There are some merge issues, partially from master but they were resolved in master. SqlInsert will be problematic merge

**Comment by SergeiPavlov** [comment]
Fixed build

> I would also add line between field and property in the SqlNodeCloneContext, I suggested this change for you to be easy to apply.

Do you mean this line? https://github.com/DataObjects-NET/dataobjects-net/pull/261/files#diff-84e02d093b91d61040b7c79479c67936f1a88cfd9cca87c9f6e3a8899566c166R13

It already presents

**Comment by alex-kulakov** [comment]
Thank you.

> It already presents

It wasn't, otherwise I wouldn't be able to make suggestion, right? I don't know what happened. Anyway, this is good that it is exists. Thank you once again

**Review by alex-kulakov — APPROVED** [review]

**Pull request merged**
